sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowGroupLimitExec.scala

```diff
@@ -0,0 +1,245 @@
+
+  override def requiredChildDistribution: Seq[Distribution] = mode match {
+    case Partial => super.requiredChildDistribution
+    case Final =>
+      if (partitionSpec.isEmpty) {
+        AllTuples :: Nil
+      } else {
+        ClusteredDistribution(partitionSpec) :: Nil
+      }
+  }
+
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] =
+    Seq(partitionSpec.map(SortOrder(_, Ascending)) ++ orderSpec)
+
+  override def outputOrdering: Seq[SortOrder] = child.outputOrdering
+
+  override def outputPartitioning: Partitioning = child.outputPartitioning
+
+  protected override def doExecute(): RDD[InternalRow] = rankLikeFunction match {
+    case _: RowNumber =>
+      child.execute().mapPartitionsInternal(
+        SimpleGroupLimitIterator(partitionSpec, output, _, limit))
+    case _: Rank =>
+      child.execute().mapPartitionsInternal(
+        RankGroupLimitIterator(partitionSpec, output, _, orderSpec, limit))
+    case _: DenseRank =>
+      child.execute().mapPartitionsInternal(
+        DenseRankGroupLimitIterator(partitionSpec, output, _, orderSpec, limit))
+  }
+
+  override protected def withNewChildInternal(newChild: SparkPlan): WindowGroupLimitExec =
+    copy(child = newChild)
+}
+
+abstract class WindowIterator extends Iterator[InternalRow] {
+
+  def partitionSpec: Seq[Expression]
+
+  def output: Seq[Attribute]
+
+  def input: Iterator[InternalRow]
+
+  def limit: Int
+
+  val grouping = UnsafeProjection.create(partitionSpec, output)
+
+  // Manage the stream and the grouping.
+  var nextRow: UnsafeRow = null
+  var nextGroup: UnsafeRow = null
+  var nextRowAvailable: Boolean = false
+  protected[this] def fetchNextRow(): Unit = {
+    nextRowAvailable = input.hasNext
+    if (nextRowAvailable) {
+      nextRow = input.next().asInstanceOf[UnsafeRow]
+      nextGroup = grouping(nextRow)
+    } else {
+      nextRow = null
+      nextGroup = null
+    }
+  }
+  fetchNextRow()
+
+  var rank = 0
+
+  // Increase the rank value.
+  def increaseRank(): Unit
+
+  // Clear the rank value.
+  def clearRank(): Unit
+
+  var bufferIterator: Iterator[InternalRow] = _
+
+  private[this] def fetchNextGroup(): Unit = {
+    clearRank()
+    bufferIterator = createGroupIterator()
+  }
+
+  override final def hasNext: Boolean =
+    (bufferIterator != null && bufferIterator.hasNext) || nextRowAvailable
+
+  override final def next(): InternalRow = {
+    // Load the next partition if we need to.
+    if ((bufferIterator == null || !bufferIterator.hasNext) && nextRowAvailable) {
+      fetchNextGroup()
+    }
+
+    if (bufferIterator.hasNext) {
+      bufferIterator.next()
+    } else {
+      throw new NoSuchElementException
+    }
+  }
+
+  private def createGroupIterator(): Iterator[InternalRow] = {
+    new Iterator[InternalRow] {
+      // Before we start to fetch new input rows, make a copy of nextGroup.
+      val currentGroup = nextGroup.copy()
+
+      def hasNext: Boolean = {
```

Comment:
note: it's allowed to call `next()` without calling `hasNext()` for java iterator. It's safe to make `hasNext` simple
```
def hasNext = nextRowAvailable && nextGroup == currentGroup
def next = ... heavy logic
```